dcrpg: start proper side chain handling, upgrades

chappjc opened the pull request:
This PR adds support for tracking blocks and transactions that are in side chains.

- [x] Add is_mainchain and is_valid to tables: blocks, transactions, vins, addresses.
- [x] Automatic postgresql table upgrades.
- [x] Change all insert and update functions to use appropriate IsMainchain and IsValid values.
- [x] Modify balance (and other relevant) queries to only consider is_mainchain = true.

For followup PR:
- Modify reorg handling to patch up the IsMainchain values.

chappjc commented:
@dmigwi There is more work here, including to ensure everything is modified to consider is_mainchain and is_valid when necessary, but I would like if you could have a look and provide feedback before I move on to the reorg parts of this.

dmigwi commented:
Sure.

dmigwi commented on db/dcrpg/internal/blockstmts.go:
```diff
@@ -50,6 +50,7 @@ const (
 		height INT4,
 		size INT4,
 		is_valid BOOLEAN,
+		is_mainchain BOOLEAN,
```
Addresses table has new column `valid_mainchain` and Blocks table has `is_mainchain`. I think using `is_mainchain` all across would ensure consistency since the variable represent the same thing but in different tables.

dmigwi replied:
No need to address this, I now understand why you used `valid_mainchain`

chappjc commented on db/dcrpg/internal/txstmts.go:
```diff
@@ -55,36 +57,52 @@ const (
+		WHERE block_hash=$2;`
+
+	UpdateTxnsValidMainchainAll = `UPDATE transactions
+		SET is_valid=b.is_valid, is_mainchain=b.is_mainchain
```
Because of the subquery into blocks?

But this will be use in two case: reorg that orphans a block, and block invalidated by stakeholders (votes in following block).  Neither cases where we can tolerate it being very slow, but not called by any user-accessible API.

chappjc replied:
Because of the subquery into blocks?

But this will be use in two case: reorg that orphans a block, and block invalidated by stakeholders (votes in following block). Neither cases where we can tolerate it being very slow, but not called by any user-accessible API.

dmigwi replied:
ok. Noted

chappjc commented on db/dcrpg/queries.go:
```diff
@@ -697,7 +702,7 @@ func DeleteDuplicateMisses(db *sql.DB) (int64, error) {
 	res, err := db.Exec(stmt, args...)
 	if err != nil {
-		return 0, fmt.Errorf(execErrPrefix + err.Error())
+		return 0, fmt.Errorf(execErrPrefix + " " + err.Error())
```
mmm, yeah, that's cleaner

chappjc replied:
mmm, yeah, that's cleaner

chappjc commented on db/dcrpg/queries.go:
```diff
@@ -1159,6 +1199,20 @@ func RetrieveTxsByBlockHash(db *sql.DB, blockHash string) (ids []uint64, txs []s
 
+// UpdateAllTxnsValidMainchain sets is_mainchain and is_valid for all
+// transactions according to their containing block.
+func UpdateAllTxnsValidMainchain(db *sql.DB) (rowsUpdated int64, err error) {
```
Agreed.  Will move to update.go.  We can consider an update/upgrade package under db/dcrpg/upgrade in the future.

chappjc replied:
Agreed. Will move to update.go. We can consider an update/upgrade package under db/dcrpg/upgrade in the future.

chappjc commented on db/dcrpg/upgrades.go:
```diff
@@ -304,6 +539,49 @@ func addNewColumnsIfNotFound(db *sql.DB) (bool, error) {
+	newColumns := []newColumn{
+		{"is_valid", "BOOLEAN", "False"},
+		{"block_time", "INT8", "False"},
+		{"value_in", "INT8", "False"},
```
good catch

chappjc commented:
I've tested the upgrade many times.  It upgrade looks like this on **testnet**:

```[INF] PSQL: Adding column is_mainchain to table blocks...
[INF] PSQL: Starting blocks table mainchain upgrade at block 000000000550c8a31ebde4238cdf5f6c9e64855d3aa9ea40b0a6eef71b77b773 (working back to genesis)
[INF] PSQL:  - 354863 rows in blocks table (sidechain/reorg upgrade) were successfully upgraded in 67.7 sec
[INF] PSQL: Adding column is_valid to table transactions...
[INF] PSQL: Adding column is_mainchain to table transactions...
[INF] PSQL: Starting transactions table mainchain upgrade...
[INF] PSQL:  - 4468353 rows in transactions table (sidechain/reorg upgrade) were successfully upgraded in 118.0 sec
[INF] PSQL: Adding column is_mainchain to table vins...
[INF] PSQL: Starting vins table mainchain upgrade...
[INF] PSQL:  - Retrieving all block hashes...
[INF] PSQL:  - Updating vins data for each transactions in every block...
[DBG] PSQL:  -- updated 1 vins for 1 of 354863 blocks
[DBG] PSQL:  -- updated 213729 vins for 10001 of 354863 blocks
[DBG] PSQL:  -- updated 492619 vins for 20001 of 354863 blocks
[DBG] PSQL:  -- updated 770795 vins for 30001 of 354863 blocks
[DBG] PSQL:  -- updated 1089940 vins for 40001 of 354863 blocks
[DBG] PSQL:  -- updated 1358289 vins for 50001 of 354863 blocks
[DBG] PSQL:  -- updated 1626488 vins for 60001 of 354863 blocks
[DBG] PSQL:  -- updated 1912679 vins for 70001 of 354863 blocks
[DBG] PSQL:  -- updated 2183717 vins for 80001 of 354863 blocks
[DBG] PSQL:  -- updated 2442177 vins for 90001 of 354863 blocks
[DBG] PSQL:  -- updated 2706658 vins for 100001 of 354863 blocks
[DBG] PSQL:  -- updated 2974538 vins for 110001 of 354863 blocks
[DBG] PSQL:  -- updated 3236623 vins for 120001 of 354863 blocks
[DBG] PSQL:  -- updated 3519468 vins for 130001 of 354863 blocks
[DBG] PSQL:  -- updated 3784287 vins for 140001 of 354863 blocks
[DBG] PSQL:  -- updated 4040284 vins for 150001 of 354863 blocks
[DBG] PSQL:  -- updated 4300844 vins for 160001 of 354863 blocks
[DBG] PSQL:  -- updated 4560536 vins for 170001 of 354863 blocks
[DBG] PSQL:  -- updated 4814112 vins for 180001 of 354863 blocks
[DBG] PSQL:  -- updated 5069455 vins for 190001 of 354863 blocks
[DBG] PSQL:  -- updated 5319792 vins for 200001 of 354863 blocks
[DBG] PSQL:  -- updated 5962063 vins for 210001 of 354863 blocks
[DBG] PSQL:  -- updated 6229022 vins for 220001 of 354863 blocks
[DBG] PSQL:  -- updated 6472656 vins for 230001 of 354863 blocks
[DBG] PSQL:  -- updated 6778381 vins for 240001 of 354863 blocks
[DBG] PSQL:  -- updated 7028858 vins for 250001 of 354863 blocks
[DBG] PSQL:  -- updated 7318845 vins for 260001 of 354863 blocks
[DBG] PSQL:  -- updated 7548246 vins for 270001 of 354863 blocks
[DBG] PSQL:  -- updated 7776095 vins for 280001 of 354863 blocks
[DBG] PSQL:  -- updated 7994711 vins for 290001 of 354863 blocks
[DBG] PSQL:  -- updated 8238719 vins for 300001 of 354863 blocks
[DBG] PSQL:  -- updated 8476481 vins for 310001 of 354863 blocks
[DBG] PSQL:  -- updated 8715918 vins for 320001 of 354863 blocks
[DBG] PSQL:  -- updated 9027462 vins for 330001 of 354863 blocks
[DBG] PSQL:  -- updated 9281892 vins for 340001 of 354863 blocks
[DBG] PSQL:  -- updated 9500038 vins for 350001 of 354863 blocks
[INF] PSQL:  - 9612800 rows in vins table (sidechain/reorg upgrade) were successfully upgraded in 801.7 sec
[INF] PSQL: Adding column valid_mainchain to table addresses...
[INF] PSQL: Starting addresses table mainchain upgrade...
[INF] PSQL: This an extremely I/O intensive operation on the database machine. It can take from 30-90 minutes.
[INF] PSQL:  - 17976681 rows in addresses table (sidechain/reorg upgrade) were successfully upgraded in 1782.1 sec
[INF] PSQL: Adding column is_mainchain to table votes...
[INF] PSQL: Starting votes table mainchain upgrade...
[INF] PSQL:  - 1710088 rows in votes table (sidechain/reorg upgrade) were successfully upgraded in 58.3 sec
[INF] PSQL: Modified the addresses table version to 3.3.0
[INF] PSQL: Modified the votes table version to 3.3.0
[INF] PSQL: Modified the vins table version to 3.3.0
[INF] PSQL: Modified the vouts table version to 3.3.0
[INF] PSQL: Modified the block_chain table version to 3.3.0
[INF] PSQL: Modified the tickets table version to 3.3.0
[INF] PSQL: Modified the misses table version to 3.3.0
[INF] PSQL: Modified the agendas table version to 3.3.0
[INF] PSQL: Modified the blocks table version to 3.3.0
[INF] PSQL: Modified the transactions table version to 3.3.0
[INF] PSQL: Table upgrades completed.
```

The vins and addresses table take the longest time.  On a fast machine, 14 minutes for vins, and 30 minutes for addresses.  Without an SSD, it will likely be 3x longer.

The necessary data columns are now present to support side chains (and clean reorg handling, whereby is_mainchain is set to false on orphaned blocks and transaction).

chappjc commented:
A quick database upgrade test can be done by copying a working DB as follows:

    create database dcrdata_reorg template dcrdata owner dcrdata;

chappjc commented:
review addressed and rebased.

dmigwi commented:
On my machine the upgrade took almost 7hrs in total. Here is the log..
```
2018-07-28 13:08:10.194 [INF] DATD: dcrdata version 3.0.0
2018-07-28 13:08:10.194 [INF] DATD: Running in full-functionality mode with PostgreSQL backend enabled.
2018-07-28 13:08:10.276 [INF] RPCC: Established connection to RPC server localhost:9109
2018-07-28 13:08:10.283 [INF] DATD: Connected to dcrd (JSON-RPC API v3.3.0) on MainNet
2018-07-28 13:08:10.288 [INF] SKDB: Loading ticket pool DB. This may take a minute...
2018-07-28 13:08:10.335 [INF] SKDB: Loading all ticket pool diffs...
2018-07-28 13:08:24.089 [INF] SKDB: Advancing ticket pool DB to tip via diffs...
2018-07-28 13:08:24.477 [INF] SKDB: Pre-populating live ticket cache and computing pool value...
2018-07-28 13:08:37.822 [INF] DATD: SQLite DB successfully opened: dcrdata.sqlt.db
2018-07-28 13:08:38.380 [INF] PSQL: Pre-loading unspent ticket info for InsertVote optimization.
2018-07-28 13:08:39.118 [INF] PSQL: Storing data for 43504 unspent tickets in cache.
2018-07-28 13:08:39.174 [INF] PSQL: Adding column is_mainchain to table blocks...
2018-07-28 13:08:47.498 [INF] PSQL: Starting blocks table mainchain upgrade at block 0000000000000000c7d8e778a65355309b4eb4b10e972d217ebe63bd73106801 (working back to genesis)
2018-07-28 13:12:33.316 [INF] PSQL:  - 255803 rows in blocks table (sidechain/reorg upgrade) were successfully upgraded in 226.5 sec
2018-07-28 13:12:33.320 [INF] PSQL: Adding column is_valid to table transactions...
2018-07-28 13:13:14.279 [INF] PSQL: Adding column is_mainchain to table transactions...
2018-07-28 13:13:58.248 [INF] PSQL: Starting transactions table mainchain upgrade...
2018-07-28 13:26:11.373 [INF] PSQL:  - 4543365 rows in transactions table (sidechain/reorg upgrade) were successfully upgraded in 733.1 sec
2018-07-28 13:26:11.404 [INF] PSQL: Adding column is_mainchain to table vins...
2018-07-28 13:26:11.436 [INF] PSQL: Starting vins table mainchain upgrade...
2018-07-28 13:26:11.436 [INF] PSQL:  - Retrieving all block hashes...
2018-07-28 13:26:12.861 [INF] PSQL:  - Updating vins data for each transactions in every block...
2018-07-28 14:57:17.748 [INF] PSQL:  - 11441405 rows in vins table (sidechain/reorg upgrade) were successfully upgraded in 5466.0 sec
2018-07-28 14:57:17.791 [INF] PSQL: Adding column valid_mainchain to table addresses...
2018-07-28 14:57:17.795 [INF] PSQL: Starting addresses table mainchain upgrade...
2018-07-28 14:57:17.795 [INF] PSQL: This an extremely I/O intensive operation on the database machine. It can take from 30-90 minutes.
2018-07-28 18:49:09.302 [INF] PSQL:  - 22219874 rows in addresses table (sidechain/reorg upgrade) were successfully upgraded in 13910.8 sec
2018-07-28 18:49:09.530 [INF] PSQL: Adding column is_mainchain to table votes...
2018-07-28 18:49:09.570 [INF] PSQL: Starting votes table mainchain upgrade...
2018-07-28 18:51:43.032 [INF] PSQL:  - 1236624 rows in votes table (sidechain/reorg upgrade) were successfully upgraded in 153.5 sec
2018-07-28 18:51:43.077 [INF] PSQL: Modified the blocks table version to 3.3.0
2018-07-28 18:51:44.188 [INF] PSQL: Modified the transactions table version to 3.3.0
2018-07-28 18:51:44.237 [INF] PSQL: Modified the block_chain table version to 3.3.0
2018-07-28 18:51:44.269 [INF] PSQL: Modified the tickets table version to 3.3.0
2018-07-28 18:51:44.297 [INF] PSQL: Modified the agendas table version to 3.3.0
2018-07-28 18:51:45.384 [INF] PSQL: Modified the vins table version to 3.3.0
2018-07-28 18:51:45.402 [INF] PSQL: Modified the vouts table version to 3.3.0
2018-07-28 18:51:46.472 [INF] PSQL: Modified the addresses table version to 3.3.0
2018-07-28 18:51:46.495 [INF] PSQL: Modified the votes table version to 3.3.0
2018-07-28 18:51:46.508 [INF] PSQL: Modified the misses table version to 3.3.0
2018-07-28 18:51:46.539 [INF] PSQL: Table upgrades completed.
```
On some of this operations/upgrades where a lot is happening in the background I think adding `Please wait ...` will give some assurances to the user that the process is still running in the background.

Also `... It can take from 30-90 minutes.` may be misleading especially if the process takes longer than 90 minutes. On my machine it took `13910.8 sec` --> `231.85 minutes`

Nice work though.

chappjc commented:
@dmigwi Wow, what a difference!  Time to upgrade to an SSD.  If you are using GCE, and have the free credits, I'd suggest creating a VM with not just a SSD persistent boot drive, but also attache a "local SSD".
